File: detection/semgrep/fixtures/missing-object-authz-check.py
# Test fixture for rule: mcp-missing-object-authz-check
#
# The guarded case uses `require_org_access`, not `requireOrgAccess`. That is
# the whole point: a Python author writes snake_case, and before the exemption
# list learned those names this correctly-guarded handler was reported as
# vulnerable. A gate wider than its defect gets switched off.


async def vuln_delete(store, token, note_id):
    session = resolve_session(store, token)
    note = store.get_note(note_id)
    # ruleid: mcp-missing-object-authz-check
    store.delete_note(note.id)
    return ok({"deleted": note.id})


async def fixed_delete_snake_case_guard(store, token, note_id):
    session = resolve_session(store, token)
    note = store.get_note(note_id)
    require_org_access(session, note)
    # ok: mcp-missing-object-authz-check
    store.delete_note(note.id)
    return ok({"deleted": note.id})


async def fixed_delete_camel_case_guard(store, token, note_id):
    """A Python codebase calling into a JS-named helper is unusual but legal."""
    session = resolve_session(store, token)
    note = store.get_note(note_id)
    requireOrgAccess(session, note)
    # ok: mcp-missing-object-authz-check
    store.delete_note(note.id)
    return ok({"deleted": note.id})

File: detection/semgrep/fixtures/missing-object-authz-check.rs
// Test fixture for rule: mcp-missing-object-authz-check-rust
//
// Rust port of missing-object-authz-check.{js,py}. The fetch and mutation
// use the real rmcp handler shape: arguments arrive wrapped in
// `Parameters<T>` and are destructured in the function signature
// (`Parameters(DeleteParams { token, note_id }): Parameters<DeleteParams>`),
// not as individual named parameters.

use rmcp::{tool, tool_router};

struct DeleteParams {
    token: String,
    note_id: String,
}

struct Note {
    id: String,
}

struct Store;

impl Store {
    fn get_note(&self, id: &str) -> Note {
        Note { id: id.to_string() }
    }
    fn delete_note(&self, id: &str) {}
}

struct Session;

fn resolve_session(store: &Store, token: &str) -> Session {
    Session
}

fn require_org_access(session: &Session, note: &Note) {}

struct NoteTools {
    store: Store,
}

#[tool_router]
impl NoteTools {
    #[tool(description = "Delete a note")]
    async fn note_delete(
        &self,
        Parameters(DeleteParams { token, note_id }): Parameters<DeleteParams>,
    ) -> String {
        let session = resolve_session(&self.store, &token);
        let note = self.store.get_note(&note_id);
        // ruleid: mcp-missing-object-authz-check-rust
        self.store.delete_note(&note.id);
        note.id
    }

    #[tool(description = "Delete a note, org-checked")]
    async fn note_delete_checked(
        &self,
        Parameters(DeleteParams { token, note_id }): Parameters<DeleteParams>,
    ) -> String {
        let session = resolve_session(&self.store, &token);
        let note = self.store.get_note(&note_id);
        require_org_access(&session, &note);
        // ok: mcp-missing-object-authz-check-rust
        self.store.delete_note(&note.id);
        note.id
    }
}
